Kibble-Zurek Mechanism with Noise Mitigation

k8culver opened the pull request:
Thanks @AndyZzzZzzZzz for your work on this feature!

I can't remember if there were additional updates we wanted to add, all I've changed is cleaning up the code a bit and fixing the tests.

@jackraymond is going to help add tests for Andy's work.

I will follow this PR with a PR updating the code structure, adding more thorough doc strings, etc, for now this is mostly just Andy's updates.

JoelPasvolsky commented on mock_kz_sampler.py:
```diff
@@ -0,0 +1,82 @@
+from dwave.system.testing import MockDWaveSampler
+
+
+class MockKibbleZurekSampler(MockDWaveSampler):
```
@k8culver, I began looking at this now but I suspect there's an update to requirements that needs to be here but is missing.
When I try to run on current requirements I get an error:
```
AttributeError: 'MockKibbleZurekSampler' object has no attribute 'mocked_parameters'
```
I guess that you need a later version of `dwave-system` and updated to 1,28 but then I get a typing error:
```
 TypeError: field job_submit_time annotation must be a type Got <module 'datetime'
```

Can you check what new requirements are needed and add them to the [requirements](https://github.com/dwave-examples/kibble-zurek/blob/main/requirements.txt) file? I can try upgrading the Ocean version but if there's a specific dependency that is now needed, it would be quicker to start with that.

k8culver replied:
I wasn't getting the same error as you on a fresh virtual env requirements install but I've upgraded the ocean requirement so let me know if that fixes the issue you found

JoelPasvolsky replied:
Looks like it's a problem for WIN OS. When I run in Unix (through codespaces) it runs as expected, but in WIN OS it fails with the following error:

```
 (kibble_zurek_3_10_11) kibble-zurek>python app.py
Traceback (most recent call last):
  File "C:\Users...kibble-zurek\app.py", line 630, in <module>
    class RunButtonClickReturn(NamedTuple):
  File "C:\Users\jpasvolsky\.pyenv\pyenv-win\versions\3.10.11\lib\typing.py", line 2279, in __new__
    nm_tpl = _make_nmtuple(typename, types.items(),
  File "C:\Users\jpasvolsky\.pyenv\pyenv-win\versions\3.10.11\lib\typing.py", line 2249, in _make_nmtuple
    types = {n: _type_check(t, f"field {n} annotation must be a type")
  File "C:\Users\jpasvolsky\.pyenv\pyenv-win\versions\3.10.11\lib\typing.py", line 2249, in <dictcomp>
    types = {n: _type_check(t, f"field {n} annotation must be a type")
  File "C:\Users\jpasvolsky\.pyenv\pyenv-win\versions\3.10.11\lib\typing.py", line 176, in _type_check
    raise TypeError(f"{msg} Got {arg!r:.100}.")
TypeError: field job_submit_time annotation must be a type Got <module 'datetime' from 'C:\\Users\\jpasvolsky\\.pyenv\\pyenv-win\\versions\\3.10.11\\lib\\datetime..
```

I thought perhaps it was a problem with updating an existing installation but the above is from a fresh installation from cloning [your branch](https://github.com/k8culver/kibble-zurek) with [this latest commit](https://github.com/dwave-examples/kibble-zurek/pull/31/commits/736f50c41243dee52006f5512aa45dc05b492869)  with its requirements for an environment of WIN 11 Pro and Python 3.10.11

thisac replied:
I had the same issue as @JoelPasvolsky and it seems to have to do with [this](https://github.com/dwave-examples/kibble-zurek/pull/31#discussion_r1919587590). Not sure why this is only an issue some times (I get it on Python 3.10, but not 3.11; haven't tested on Windows).

k8culver replied:
Updated those types, let me know if that fixed it

JoelPasvolsky replied:
yes, that fixes the problem 👍

thisac reviewed (commented):
Some high-level comments (still have to review most of the code):
* The logo and the tab-buttons aren't aligned with the edge and jump around quite a bit when resizing the window. It'd look better if they where at a constant distance from the edges, I think.
* When running the ZNE simulation, curve fitting seems to happen after a couple of runs with different coupling strengths. This doesn't seem to work as it should. If the fit fails, there's a popup and everything is temporarily gone. If the fit finishes and a curve appears, the scaling seems to be off. It also disappears (being replaced with the sample data points) if switching view.
* The original demo has no default value for QPU, causing an error (instead of a user-friendly warning) if a run is started without a chosen QPU.

JoelPasvolsky commented on helpers/tooltips.py:
```diff
@@ -12,34 +12,26 @@
-f"""Coupling strength, J, between spins in the ring. 
-Range of -2 (ferromagnetic) to +1 (anti-ferromagnetic).
+tool_tips_demo2 = {
+    "anneal_duration": f"""Duration of the quantum anneal. Range of 5 to 320 nanoseconds.""",
```
> Not really to do with this update, but this tooltip doesn't show for me (for either tab).

It shows in the current version in main but not in this PR:
![image](https://github.com/user-attachments/assets/5bfb347b-c59d-456f-a2b6-0f4644198753)

k8culver replied:
Fixed: https://github.com/dwave-examples/kibble-zurek/pull/31/commits/22a690ed4989ade6da97a270f2dfa9d921c87b1e

k8culver replied:
I meant to rename all the html classes but I'm going to save it for a different PR, but this is good for me to remember to change the tooltips too if I do that

k8culver commented:
> * The logo and the tab-buttons aren't aligned with the edge and jump around quite a bit when resizing the window. It'd look better if they where at a constant distance from the edges, I think.

The break point jumping behaviour is standard practice for web development that Dash has built in with their Container component. But I agree that the sides should be lined up so I'll remove the Container here.

k8culver commented:
> * The original demo has no default value for QPU, causing an error (instead of a user-friendly warning) if a run is started without a chosen QPU.

Yeah I agree, I wonder if @JoelPasvolsky has a suggestion on which would make a good default?

k8culver commented on README.md:
```diff
@@ -118,6 +118,19 @@ Your development environment should be configured to
 Leap account [here](https://docs.dwavesys.com/docs/latest/doc_leap_dev_env.html).  
 
+The default configuration uses `DWaveSampler` with specific models accessed through the Leap API.
+To run experiments using `MockDKibbleZurekSampler` locally, set the environment variable in your
```
Updated: https://github.com/dwave-examples/kibble-zurek/pull/31/commits/8a88a8408cb38288ed0aab1fa6e6cca24ef72b82

k8culver commented on README.md:
```diff
@@ -190,6 +206,22 @@ Note that as you increase the anneal time, you move from the coherent regime
+Another feature showcased in this demo is the result achieved in Quantum Error Mitigation.
+In [this paper](https://arxiv.org/abs/2311.01306), we demonstrate a practical implementation
+of zero-noise extrapolation as a method of quantum error mitigation specifically used for quantum
+annealing.
```
Updated: https://github.com/dwave-examples/kibble-zurek/pull/31/commits/8a88a8408cb38288ed0aab1fa6e6cca24ef72b82

k8culver commented:
> * When running the ZNE simulation, curve fitting seems to happen after a couple of runs with different coupling strengths. This doesn't seem to work as it should. If the fit fails, there's a popup and everything is temporarily gone. If the fit finishes and a curve appears, the scaling seems to be off. It also disappears (being replaced with the sample data points) if switching view

@thisac I think I fixed this bug but let me know if you're still seeing it

k8culver commented:
@JoelPasvolsky @jackraymond @thisac I've made quite a few updates but I believe we're ready for another round of reviews.

JoelPasvolsky reviewed (commented):
This is a vast improvement, @k8culver, wow.  

Just from playing with it for a bit here are some comments and questions:

1. It's unclear to me how the ZNE panel's lower graph can have a predicted line for several values of J: the kink density is dependent to some extent on J so each plotted coupling strength would have a slightly different slope, yet the plot has a single predicted band. I think that might have to be removed, no?

![image](https://github.com/user-attachments/assets/ad84ea00-cbe7-46a9-8d4e-f26000edc2c1)


2. Panel names: perhaps "Kibble-Zurek Simulation" and "Zero-Noise Extrapolation"?
3. Plot names: currently just describing axes and for the lower one using two different names (quench and anneal). Can these be more descriptive of what they are doing? The top one is extrapolating the zero-noise density and the bottom one is displaying the measured and extrapolated kink densities
4.  Left panel introduction: currently explains the science but might be more useful to tell the user how to run the demo (it's more complicated now and needs a few steps to get useful results) and put the science explanation in the readme.
5. Needs rebasing (currently does not have the adv7 cached files)
6. I'm getting debug prints in my terminal

![image](https://github.com/user-attachments/assets/849fce7b-4158-46e3-a080-f3463bcea6fc)

I'll also go over the readme. If I have time I'll look over the code changes but I might not and have to leave that to T and J.

k8culver commented on README.md:
```diff
@@ -190,6 +194,18 @@ Note that as you increase the anneal time, you move from the coherent regime
+as described in this [paper](https://arxiv.org/abs/2311.01306).
+
+A fitting function—quadratic for the Advantage solver and a multi-polynomial for the
+MockDwaveSampler can be used to calculate the theoretical zero-noise point for various coupling
```
Good catch I missed this, fixed here: https://github.com/dwave-examples/kibble-zurek/pull/31/commits/d18c3187168efa9e386db0c12ad18b73ba7ca8d0

jackraymond replied:
Remove the MockSampler and all references thereto. This was only introduced for debugging, and the phenomena is actually quite misleading (thence even the need for a different fit type).

k8culver commented on README.md:
```diff
@@ -190,6 +194,18 @@ Note that as you increase the anneal time, you move from the coherent regime
+## <a name="ZNE"></a> Zero-Noise Extrapolation
+
+Zero-Noise Extrapolation (ZNE) is a quantum error mitigation method used for quantum annealing
+as described in this [paper](https://arxiv.org/abs/2311.01306).
```
Updated to a citation: https://github.com/dwave-examples/kibble-zurek/pull/31/commits/d18c3187168efa9e386db0c12ad18b73ba7ca8d0

jackraymond replied:
We'll need to add the journal reference when published.

k8culver commented:
@JoelPasvolsky I think I've addressed everything. I'm not sure how to proceed with 1) and 4) based on @jackraymond's feedback

> 1. It's unclear to me how the ZNE panel's lower graph can have a predicted line for several values of J: the kink density is dependent to some extent on J so each plotted coupling strength would have a slightly different slope, yet the plot has a single predicted band. I think that might have to be removed, no?

Will defer to @jackraymond on this as he understands the problem better.

> 2. Panel names: perhaps "Kibble-Zurek Simulation" and "Zero-Noise Extrapolation"?

Updated here: https://github.com/dwave-examples/kibble-zurek/pull/31/commits/d18c3187168efa9e386db0c12ad18b73ba7ca8d0

> 3. Plot names: currently just describing axes and for the lower one using two different names (quench and anneal). Can these be more descriptive of what they are doing? The top one is extrapolating the zero-noise density and the bottom one is displaying the measured and extrapolated kink densities

Updated to "Extrapolating Zero-Noise Density" and "Measured and Extrapolated Kink Densities", let me know if those make sense. Updated here: https://github.com/dwave-examples/kibble-zurek/pull/31/commits/d18c3187168efa9e386db0c12ad18b73ba7ca8d0

> 4. Left panel introduction: currently explains the science but might be more useful to tell the user how to run the demo (it's more complicated now and needs a few steps to get useful results) and put the science explanation in the readme.

I am planning on doing a follow up PR to have the autoplay option. Would this solve this problem?

> 5. Needs rebasing (currently does not have the adv7 cached files)

These changes will show once merged. If you would like this for testing I can pull the changes in.

> 6. The button to turn off tool tips seems to have gone missing. That is useful for people giving this demo to an audience.

I removed it as it cluttered the UI and distracted from the problem settings, I've added the functionality back but as a setting in the `demo_configs.py` file, which will let presenters hide the tooltips before running the problem. Updated here: https://github.com/dwave-examples/kibble-zurek/pull/31/commits/6a3a94e9e2c0902a316d79d1590bf7e6c27a3405

> 7. I'm getting debug prints in my terminal

These can be hidden by setting `DEBUG` to `True` in `demo_configs.py` which will make the debug info appear in the debug menu (blue bottom right circle) in the UI.

jackraymond commented:
Minor requests.
1. X-axis on top plot, "Noise ratio (-1.8/J)" -> "Noise ratio (t_{programmed}/t_{target})"
2. "Extrapolating Zero-Noise Density" -> "Zero-noise extrapolation of kink density"
3. "Measured and extrapolated kink density" -> "Measured and extrapolated kink density for target J=-1.8 chain"
4. "Quench/Anneal Duration [ns]" -> "Target Quench Duration [ns]"
5. X-axis bottom plot "Target Quench Duration [ns]"
6. "Owing to thermal noise, coupled chains depart from closed system dynamics (Kibble-Zurek power law 1/2 scaling) for longer quench duration. Experiments at smaller coupling strength allow for the modelling of higher noise environments. We can model at a range of coupilng strengths (noise levels), and extrapolate towards the noise-free regime; thereby improving agreement with theory (closed system dynamics) to larger quench durations. "
To incorporate Joel's feedback:
"Statistics of a (target)  J=-1.8 chain at quench duration $t_{target}$, can be inferred by running at weaker coupling and longer quench duration ($t_{programmed}$). Longer programmed times (at weaker coupling) are subject to more noise. Collecting data at several noise levels an extrapolation to a denoised result is possible.
At short target time scales there is weak environmental coupling, denoising has little impact.
At long target time scales there is strong environmental coupling, denoising improves agreement with Kibble-Zurek theory.
"

k8culver commented:
> Some ideas, second opinion would be helpful:
> 
> 1. X-axis on top plot, "Noise ratio (-1.8/J)" -> "Noise ratio (t_{programmed}/t_{target})"
> 2. "Extrapolating Zero-Noise Density" -> "Zero-noise extrapolation of kink density"
> 3. "Quench/Anneal Duration [ns]" -> "Target Quench Duration [ns]"
> 4. X-axis bottom plot "Target Quench Duration [ns]"
> 5. "Owing to thermal noise, coupled chains depart from closed system dynamics (Kibble-Zurek power law 1/2 scaling) for longer quench duration. Experiments at smaller coupling strength allow for the modelling of higher noise environments. We can model at a range of coupilng strengths (noise levels), and extrapolate towards the noise-free regime; thereby improving agreement with theory (closed system dynamics) to larger quench durations. "
>    To incorporate Joel's feedback:
>    "Statistics of a (target)  J=-1.8 chain at quench duration t_target, can be inferred by running at weaker coupling and longer quench duration (t_programmed).
>    Longer programmed times (at weaker coupling) are subject to more noise. Collecting data at several noise levels an extrapolation to a denoised result is possible.
>    At short target time scales there is weak environmental coupling, denoising has little impact.
>    At long target time scales there is strong environmental coupling, denoising improves agreement with Kibble-Zurek theory."

@jackraymond made these changes here: https://github.com/dwave-examples/kibble-zurek/pull/31/commits/d9c9374a3f71ff2abbf39e101f983afd6832ccc1

thisac commented on demo_configs.py:
```diff
@@ -0,0 +1,42 @@
+DESCRIPTION = """\
+Use a quantum computer to simulate the formation of topological defects in a 1D ring 
+of spins undergoing a phase transition, described by the Kibble-Zurek mechanism.
+"""
+MAIN_HEADER_NM = "Zero-Noise Extrapolation"
```
(minor) Separating these makes it a bit cleaner IMO.
```suggestion
"""

# config settings for ZNE tab
MAIN_HEADER_NM = "Zero-Noise Extrapolation"
```

thisac commented on helpers/plots.py:
```diff
@@ -164,182 +275,408 @@ def plot_kink_densities_bg(display, time_range, coupling_strength, schedule_name
+        plotly.graph_objs.Figure: A Plotly figure object.
+    """
+    fig_layout = go.Layout(
+        xaxis3=dict(title="<b>Noise ratio (t_{programmed}/t_{target})</b>", type="linear", range=[0, 3]),
```
```suggestion
        xaxis3=dict(title="<b>Noise ratio (t_programmed/t_target)</b>", type="linear", range=[0, 3]),
```

thisac commented on helpers/plots.py:
```diff
@@ -164,182 +275,408 @@ def plot_kink_densities_bg(display, time_range, coupling_strength, schedule_name
     Args:
+        time_range (list of float): A list containing the minimum and maximum
+            quench times [min_quench_time, max_quench_time] in nanoseconds.
+        n: TODO
```
TODO?

k8culver replied:
Fixed: https://github.com/dwave-examples/kibble-zurek/pull/31/commits/a8e6d14c08754976a1ded5e1300c60751df154ad

thisac reviewed (approved):
Looks really great. Minor, but I noticed that it's still possible to break it by setting the quench duration to less than 5 (not by using the arrows). Apart from that and a few minor comments, from a code and function/UX perspective, LGTM! 

I'll let @jackraymond determine whether the scientific details all are good.

k8culver commented:
> Looks really great. Minor, but I noticed that it's still possible to break it by setting the quench duration to less than 5 (not by using the arrows). Apart from that and a few minor comments, from a code and function/UX perspective, LGTM!

I'm going to consider this update as scope creep for this PR since it's the original demo that has the issue not the update. I'll add this as an issue for a future fix.

@JoelPasvolsky @jackraymond if you two are happy with everything I think we're good to go

jackraymond commented on helpers/qa.py:
```diff
@@ -12,126 +12,230 @@
+    }
+
+
+def fitted_function(xdata, ydata, method="polynomial", degree=1):
```
The only method used by the tutorial is 'pure_quadratic'; where degree is not used. With this in mind we could delete the method and degree variables and just keep the single relevant branch.

Having done this we can add a simple test using f(x) = 1+x^2. xdata = [0,2], ydata=[1,5] and then check that when we call the return value with argument x=1 it gives result 2 up to small error (like 1e-8).

k8culver replied:
Added test: https://github.com/dwave-examples/kibble-zurek/pull/31/commits/69b62d739f634f2092e5a6f15ad218869eaa6393

jackraymond reviewed (changes requested):
We can simplify and test the fitted function. I'd like to delete most of the code in fitted_function(), since it was created to support non-Leap development/testing that is no longer relevant - I suggested a simple test for the remaining part.
 I can do this on Monday if @k8culver is busy.

doi:10.1038/s41534-025-00977-3
Paper is going to be published (I think with the above DOI) in a few days. So we can replace references to the arxiv version with this new version.

I suggested t_programmed and t_fitted; Might look more standard to have t with either programmed or fitted as subscripts. It's a technicality.

target time scales -> target quench durations (for greater consistency with other labels)
